server/test/unit/com/thoughtworks/go/server/materials/MaterialUpdateServiceTest.java

```diff
@@ -112,25 +122,111 @@ public void teardown() throws Exception {
+    }
+
+    @Test
+    public void shouldAllowConcurrentUpdatesForNonAutoUpdateMaterials_updateMaterial() throws Exception {
+        ScmMaterial material = mock(ScmMaterial.class);
+        when(material.isAutoUpdate()).thenReturn(false);
+        MaterialUpdateMessage message = new MaterialUpdateMessage(material, 0);
+        doNothing().when(queue).post(message);
+
+        assertTrue(service.updateMaterial(material)); //prune inprogress queue to have this material in it
+        assertTrue(service.updateMaterial(material)); // immediately notify another check-in
+
+        verify(queue, times(2)).post(message);
+        verify(material).isAutoUpdate();
+    }
+
+    @Test
+    public void shouldNotAllowConcurrentUpdatesForAutoUpdateConfigMaterials_updateMaterial() throws Exception {
+        ScmMaterial material = mock(ScmMaterial.class);
+        when(material.isAutoUpdate()).thenReturn(true);
+        when(material.getFingerprint()).thenReturn("fingerprint");
+        when(watchList.hasConfigRepoWithFingerprint("fingerprint")).thenReturn(true);
+        MaterialUpdateMessage message = new MaterialUpdateMessage(material, 0);
+        doNothing().when(configQueue).post(message);
+
+        assertTrue(service.updateMaterial(material)); //prune inprogress queue to have this material in it
+        assertFalse(service.updateMaterial(material)); // immediately notify another check-in
+
+        verify(configQueue, times(1)).post(message);
+        verify(material).isAutoUpdate();
+    }
+
+    @Test
+    public void shouldNotAllowConcurrentUpdatesForAutoUpdateMaterials_updateMaterial() throws Exception {
+        ScmMaterial material = mock(ScmMaterial.class);
+        when(material.isAutoUpdate()).thenReturn(true);
+        MaterialUpdateMessage message = new MaterialUpdateMessage(material, 0);
+        doNothing().when(queue).post(message);
+
+        assertTrue(service.updateMaterial(material)); //prune inprogress queue to have this material in it
+        assertFalse(service.updateMaterial(material)); // immediately notify another check-in
+
+        verify(queue, times(1)).post(message);
+        verify(material).isAutoUpdate();
+    }
+
+    @Test
+    public void shouldAllowPostCommitNotificationsToPassThroughToTheQueue_WhenTheSameMaterialIsNotCurrentlyInProgressAndMaterialIsAutoUpdateTrue() throws Exception {
+        ScmMaterial material = mock(ScmMaterial.class);
+        when(material.isAutoUpdate()).thenReturn(true);
+        MaterialUpdateMessage message = new MaterialUpdateMessage(material, 0);
+        doNothing().when(queue).post(message);
+
+        service.updateMaterial(material); // first call to the method
+
+        verify(queue, times(1)).post(message);
+        verify(material, never()).isAutoUpdate();
+    }
+
+    @Test
+    public void shouldAllowPostCommitNotificationsToPassThroughToTheQueue_WhenTheSameMaterialIsNotCurrentlyInProgressAndMaterialIsAutoUpdateFalse() throws Exception {
```

Comment:
Now that we have a "retry" queue concept, if might be worth keeping these post commit notifications in a similar queue as well. Not for this PR/story. But still, think about it. Right now, those notifications are discarded.